save/src/lib.rs

```diff
@@ -46,18 +46,66 @@ impl Save {
         &self.meta
     }
 
-    pub fn get(&self, node_id: u128) -> Result<Option<Node>, GetError> {
+    pub fn read(&self) -> Result<ReaderGuard<'_>, redb::Error> {
         let tx = self.db.begin_read()?;
-        let nodes = tx.open_table(NODE_TABLE)?;
-        let Some(node) = nodes.get(&node_id)? else { return Ok(None); };
-        Ok(Some(Node::decode(node.value())?))
+        Ok(ReaderGuard { tx })
     }
 
-    pub fn put(&self, node_id: u128, node_state: Node) -> Result<(), redb::Error> {
-        let msg = node_state.encode_to_vec();
+    pub fn write(&mut self) -> Result<WriterGuard<'_>, redb::Error> {
         let tx = self.db.begin_write()?;
-        let mut nodes = tx.open_table(NODE_TABLE)?;
-        nodes.insert(node_id, &*msg)?;
+        Ok(WriterGuard { tx })
+    }
+}
+
+pub struct ReaderGuard<'a> {
+    tx: redb::ReadTransaction<'a>,
+}
+
+impl ReaderGuard<'_> {
+    pub fn get(&self) -> Result<Reader<'_>, redb::Error> {
+        let table = self.tx.open_table(NODE_TABLE)?;
+        Ok(Reader { table })
+    }
+}
+
+pub struct Reader<'a> {
+    table: redb::ReadOnlyTable<'a, u128, &'static [u8]>,
```

Comment:
I'm surprised the lifetime is `'static` here. It seems suspicious, as I'm not sure how ReDb can possibly expose a `'static` array reference without keeping it in memory forever.

Reply:
Good instincts. In this case the `RedbValue` trait has a GAT that swaps out `'static` with the appropriate lifetime when accessing a value, so `'static` is really just a placeholder. The `&[u8]` used in e.g. the `const` table definitions is inferred to be the same.

Reply:
Interesting. Thanks for the extra context. This definitely still confuses me a bit, but I should be able to learn this better once I find the time to look into it further.